python/pyarrow/src/arrow/python/python_to_arrow.cc

```diff
@@ -762,7 +762,7 @@ class PyListConverter : public ListConverter<T, PyConverter, PyConverterTrait> {
       RETURN_NOT_OK(AppendSequence(value));
     } else if (PySet_Check(value) || (Py_TYPE(value) == &PyDictValues_Type)) {
       RETURN_NOT_OK(AppendIterable(value));
-    } else if (PyDict_Check(value) && this->options_.type->id() == Type::MAP) {
+    } else if (PyDict_Check(value) && this->type()->id() == Type::MAP) {
```

Comment:
It looks like `options_` isn't updated/set as expected in `MakeConverter`

So, when there's a nested map (e.g. list of map):
1. `type_` is updated to be the list's value_type, but
2. `option_` is passed along unchanged and still has its `type_` set as the list

That causes this check to fail unexpected, since the pydict is seen as having `options_.type->id() == Type::LIST` instead of `type()->id() == Type::MAP`

I am unsure what the expected behavior of `options_` is, but this check certainly is meant to check the current type.

Reply:
Agreed. `MakeConverter` doesn't seem to expect that options is specific to the type in question; it's templated over it. This change seems correct to me. Thanks for catching it. 👍